### PR TITLE
Feature/android touchless sms receive

### PR DESCRIPTION
**Pull request by drilonreqica**
Implemented SMS Receiving and Parsing.

### CODE REVIEW

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        }
+    };
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+                                           @NonNull String permissions[],
+                                           @NonNull int[] grantResults) {
+        switch (requestCode) {
+            case SmsTool.REQUEST_CODE_ASK_PERMISSIONS: {
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: please remove braces from case statement

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                // If request is cancelled, the result arrays are empty.
+                if (grantResults.length <= 0
+                        || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
+                    Toast.makeText(this, "App cannot work without RECEIVE_SMS permission!",
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: put into string resource --> error_permission_receive_sms

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import android.Manifest;
+import android.app.Activity;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.os.Build;
+import android.support.annotation.NonNull;
+import android.support.annotation.RequiresApi;
+import android.support.v4.content.ContextCompat;
+
+public class SmsTool {
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java]: make class final and convert to util class with private constructor and 

`public static void requestSMSPermission(@Nonnull Activity activity)`

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        SmsTool smsTool = new SmsTool(this);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: method is annotated with
`@RequiresApi(api = Build.VERSION_CODES.M)`so the sdk version check should be redundant

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            smsTool.requestSMSPermission();
+        }
+
+        initViews();
+
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: remove empty line

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        registerReceiver();
+    }
+
+    @Override
+    protected void onPause() {
+        localBroadcastManager.unregisterReceiver(broadcastReceiver);
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: either also use a unregisterReceiver method or don't use a registerReceiver method to stay consistent in your code

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        setContentView(R.layout.activity_main);
+
+        SmsTool smsTool = new SmsTool(this);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            smsTool.requestSMSPermission();
+        }
+
+        initViews();
+
+        registerReceiver();
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: remove registerReceiver call --> already called in onResume

File: secure-mobile-push/smp-android-client/app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1,3 +1,6 @@
 <resources>
     <string name="app_name">Secure Mobile Push</string>
+    <string name="send_a_sms_for_app_to_read">Send a SMS for the App to Read</string>
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/app/src/main/res/values/strings.xml]: please rename to 

- text_sms_send
- text_sms_sender_number
- text_sms_message

and change string to "Send a SMS to the application"

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import de.adorsys.android.securemobilepush.BuildConfig;
+
+public class SmsReceiver extends BroadcastReceiver {
+    public static final String INTENT_FILTER_SMS = "intentFilterSms";
+    public static final String KEY_SMS_SENDER = "smsSender";
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: string should have the same value as the variable --> key_sms_sender

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import de.adorsys.android.securemobilepush.BuildConfig;
+
+public class SmsReceiver extends BroadcastReceiver {
+    public static final String INTENT_FILTER_SMS = "intentFilterSms";
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: please rename to INTENT_ACTION_SMS and value to "intent_action_sms"

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    public static final String KEY_SMS_SENDER = "smsSender";
+    public static final String KEY_SMS_MESSAGE = "smsMessage";
+
+    private static final String INTENT_SMS_ACTION = "android.provider.Telephony.SMS_RECEIVED";
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        if (intent.getAction().equals(INTENT_SMS_ACTION)) {
+            List smsSenderNumbers =
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: no type specified for List --> probably `List<String>`

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                    new ArrayList<>(Arrays.asList(BuildConfig.SMS_SENDER_NUMBER));
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: can't you just write 
`List<String> smsSenderNumbers = Arrays.asList(BuildConfig.SMS_SENDER_NUMBERS);`?

File: secure-mobile-push/smp-android-client/securemobilepush/build.gradle [hunk 13]
```diff
@@ -5,14 +5,25 @@ android {
...
-        versionName "1.0"
+        versionName "1.0.0"
 
         testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+    }
 
+    defaultConfig {
+        buildConfigField "String", "BEGIN_INDEX", "\"BEGIN-MESSAGE\""
+        buildConfigField "String", "END_INDEX", "\"END-MESSAGE\""
+        buildConfigField "String[]", "SMS_SENDER_NUMBER",
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/build.gradle]: as this is now an array, please rename field to `SMS_SENDER_NUMBERS`

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+            Bundle bundle = intent.getExtras();
+            SmsMessage[] smsMessages;
+            String messageFrom;
+            if (bundle != null) {
+                try {
+                    Object[] pdus = (Object[]) bundle.get("pdus");
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: either comment what's going on here or extract string to static member variable which speaking name

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+                    if (pdus != null) {
+                        smsMessages = new SmsMessage[pdus.length];
+                        for (int i = 0; i < smsMessages.length; i++) {
+                            String format = bundle.getString("format");
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: create string inside `if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {`

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+                } catch (Exception e) {
+                    if (BuildConfig.DEBUG) {
+                        Log.d(SmsReceiver.class.getName(), e.getMessage());
+                    }
+                }
+            }
+        }
+    }
+
+    private String getSmsCode(@NonNull String message) {
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: can return value be null ? @Nullable : @NonNull

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+                        for (int i = 0; i < smsMessages.length; i++) {
+                            String format = bundle.getString("format");
+                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                                smsMessages[i] = SmsMessage.createFromPdu((byte[]) pdus[i], format);
+                            } else {
+                                smsMessages[i] = SmsMessage.createFromPdu((byte[]) pdus[i]);
+                            }
+                            messageFrom = smsMessages[i].getOriginatingAddress();
+                            if (smsSenderNumbers.contains(messageFrom)) {
+                                String messageBody = getSmsCode(smsMessages[i].getMessageBody());
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: I would extract this code block into a new method called sendBroadcast or send

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+                                Intent broadcastIntent = new Intent(INTENT_FILTER_SMS);
+                                broadcastIntent.putExtra(KEY_SMS_SENDER, messageFrom);
+                                broadcastIntent.putExtra(KEY_SMS_MESSAGE, messageBody);
+                                LocalBroadcastManager.getInstance(context).sendBroadcast(broadcastIntent);
+                            }
+                        }
+                    }
+                } catch (Exception e) {
+                    if (BuildConfig.DEBUG) {
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: I would also send the broadcast in this case but with empty payload or extra error flag to be able to react in the UI to errors, e.g.: stop showing progress bar

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+        this.context = context;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    public void requestSMSPermission() {
+        Activity activity = (Activity) context;
+        final String permission = Manifest.permission.RECEIVE_SMS;
+        int hasSpecificPermission = ContextCompat.checkSelfPermission(context, permission);
+        if (hasSpecificPermission != PackageManager.PERMISSION_GRANTED) {
+            if (!activity.shouldShowRequestPermissionRationale(permission)) {
```
Comment (itsmortoncornelius) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java]: if statements can be combined

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 14 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/fce7b9891678954f6e4af146afdbf15b98bc418e

File: secure-mobile-push/smp-android-client/securemobilepush/build.gradle (hunk 13 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/build.gradle]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/c543cd4e7898e1a8cc65474c087c0f69515c2a15

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 16 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/cf309436bc8adf76dfc227e12e77d62ce79d3a86

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 18 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/3639a8daafabdf2f833b2d5997ab5854bbd4ca98

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 17 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/bc533c9ebe202482e9109dfc6c33f3afd226fe53

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 15 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in Commit: https://github.com/adorsys/secure-key-storage/commit/20bfccdd5ef5691f3ce3663574c3d489365756b3

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 12 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/1d72b1781845022eae7162cf891341d3018a5971

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 11 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/1d72b1781845022eae7162cf891341d3018a5971

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 9 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/956ba7e93c71c09d0ecf588c93883c4cacb8740a

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java (hunk 10 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsReceiver.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/956ba7e93c71c09d0ecf588c93883c4cacb8740a

File: secure-mobile-push/smp-android-client/app/src/main/res/values/strings.xml (hunk 8 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/app/src/main/res/values/strings.xml]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/87a2a2c02f2d278a60065555e93e83c276326ff5

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java (hunk 19 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/fd9ee5c6e8926d9e0ff95c5c30b8531ff5143d8b

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java (hunk 6 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/345504beb38fd68a781658f6e2dea1cce1bc4b8d

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java (hunk 7 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/120ecffd53394139c5b38985c83573e2f4a96bb2

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java (hunk 5 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/912989927f13a9928d970dfb92ae0e660be10135

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java (hunk 2 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/1fdbf69ba5439bc8bca27f4d2051ea4e67592daf

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java (hunk 1 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/c2a7bede2ed43af99ec8c632ea9e6a03bbfbca8a

File: secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java (hunk 4 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/app/src/main/java/de/adorsys/android/securemobilepushtest/MainActivity.java]: @RequiresApi requires user to check the actual API version, unlike @TargetApi.
This behavior can be compared to @Nullable & @NonNull/

File: secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java (hunk 3 above)
Comment (drilonreqica) [inline secure-mobile-push/smp-android-client/securemobilepush/src/main/java/de/adorsys/android/securemobilepush/sms/SmsTool.java]: Fixed in commit: https://github.com/adorsys/secure-key-storage/commit/761c856002a1a3f6393ea09c990158803e09aa51

**Pull request merged**